

```{r, htmlcap="Mutational spectrum specific S>NS conversion factors differ between four orders of magnitude. The NS>E conversion factors are all within the same order magnitude", warning=FALSE}
# compute_rates(vn1 = 'var_s', vn2 = 'var_ns', rooney_list)

pdf(file.path(img_loc, 'rooney_analysis_cf_dists.pdf'))
layout(matrix(c(1,2), nrow = 2))
hist(log10(rooney_cfs[[1]][[1]]),
     main = "S>NS rates per mutational spectrum",
     xlab = "log10(non-synonymous mutations / synonymous mutations)", col = 'navyblue')
hist(rooney_cfs[[1]][[2]], main = "NS>E rates per mutational spectrum",
     xlab = "neo-epitopes / non-synonymous mutations",
     col = 'navyblue')
dev.off()
```


We computed the silent-to-non-synonymous conversion rate by summing all
non-synonymous and synonymous variants in all donors on a per mutation spectrum
basis and dividing the former sum by the latter. During computation of mutation
spectrum specific conversion factors, some conversion factors couldn't be
computed as either the reference codon codes for a stop codon (A(A|C)>AT) or the
mutant codon does. For these following mutational spectra, no synonymous
mutations were recorded although non-synonymous were recorded.  Note that none
of the mutations analyzed here was indicated to be a stop codon by *VarContext*.
Additionally, one spectrum is strongly overrepresented in the non-silent group
(i.e. more likely to yield non-silent than silent mutations).  These weirdly
behaving spectra could be blowing up the non-syn predictions (and thereby the
predicted epitope amounts).  Remarkably, all odd behaving patterns are flanked
by an A on the 5' side and a T on the 3' side. Below a table of all such
patterns is shown.  One could perhaps remedy these problems by taking the codon
positions of these mutations into consideration, but I can't find a notion of
that in Rooney et al.  (2015).

> Should I look further into normalizing the conversion factors here? Did you also
> observe such out of proportion values and if yes did you choose to exclude them from
> analysis?

from `https://en.wikipedia.org/wiki/Stop_codon`
In the standard genetic code, there are three different stop codons:
in RNA/DNA:
UAG/TAG ("amber")
UAA/TAA ("ochre")
UGA/TGA ("opal" or "umber")

    pos 1 of consensus sequence 
    *TA
    *AA
    *GA
    pos 3 of consensus sequence
    AG*
    AA*
    GA*


```{r, eval = F, warning=FALSE, echo=FALSE}
agg_dvc <- w_readRDS('agg_dvc')

df <- unique(agg_dvc[seq_effect == 'stop_gained' &
                     stringr::str_length(ref_allele) == 1 & 
                     stringr::str_length(alt_allele) == 1, 
             .(strand, chromosome, start_position, end_position,
               ref_allele, alt_allele, 
               codon_germline, codon_tumor,
               'ensembl_query' = sprintf('%s:%s-%s', chromosome, as.character(start_position-2), 
                       as.character(start_position+2)), 
               mut_context, codon_pos)])[order(mut_context)]


reverse_complement_mut_context <- function(vec) {
  comp_string <- c('A' = 'T', 'T' = 'A', 'G' = 'C', 'C' = 'G')
  sapply(vec, function(i) {
    j5 <- comp_string[substring(i, 5, 5)]
    j4 <- comp_string[substring(i, 4, 4)]
    j2 <- comp_string[substring(i, 2, 2)]
    j1 <- comp_string[substring(i, 1, 1)]
    substr(i, 1, 1) <- as.character(j5)
    substr(i, 2, 2) <- as.character(j2)
    substr(i, 4, 4) <- as.character(j4)
    substr(i, 5, 5) <- as.character(j1)
    i
  })
}

invisible(df[strand == "-1", mut_context := reverse_complement_mut_context(mut_context)])

# head(df)
# df <- unique(df[, .(mut_context, codon_pos)])[order(codon_pos, mut_context)]
df[codon_pos == 2]
```


```{r, warning=FALSE, echo=FALSE}
## Everything going in to the ATC codon (stop)
rooney_cfs[['StNS']] %>% { .[. == Inf] }
rooney_cfs[['StNS']][grep("T(A|T|C|G)>AG", names(rooney_cfs[[1]]), value = T)]
rooney_cfs[['StNS']][grep("TG>(A|T|C|G)A", names(rooney_cfs[[1]]), value = T)]
rooney_cfs[['StNS']][grep("T(A|T|C|G)>AA", names(rooney_cfs[[1]]), value = T)]
rooney_cfs[['StNS']][grep("TA>(A|T|C|G)A", names(rooney_cfs[[1]]), value = T)]
rooney_cfs[['StNS']][grep("A(A|T|C|G)>TC", names(rooney_cfs[[1]]), value = T)]
rooney_cfs[['StNS']][grep("AT>(A|T|C|G)C", names(rooney_cfs[[1]]), value = T)]
rooney_cfs[['StNS']] %>% { .[is.na(.) | . >= 100] } %>%
  { data.table(mut_context = names(.), 'S>NS' = .) } %>%
  { .[order(mut_context)] } %>%
  { print(knitr::kable(.)) }
```
