---
title: "Rooney-inspired immune editing"
author: "Maarten Slagter"
output:
  html_document:
    toc: true
    tocdepth: 2
    toc_float: false
    number_sections: true
    fig_caption: true
---

<style>
img { max-width: none; }
</style>


```{r, init, include = F, warning=FALSE, echo=FALSE}
knitr::opts_chunk$set(cache = T, fig.height = 5, fig.width = 1.66 * 5)
knitr::opts_chunk$set(include = TRUE, warning = FALSE, message = FALSE, echo = FALSE)
knitr::knit_hooks$set(htmlcap = function(before, options, envir) {
  if(!before) {
    paste0('<p class="caption", align="center"><i>', options$htmlcap, "</i></p>")
  }
})
blacklist_projs <- c('CLLE-ES', 'CoMMpass')
options(digits = 3)
pacman::p_load(dtplyr)
pacman::p_load(ggplot2)
pacman::p_load(grid)
```


## Analysis 1 (see supplements)


```{r, rooney_real_all, eval = F, cache = F, warning=FALSE}
R_file <- file.path(root_folder, 'maarten-analyses', 
                    'immune_editing', 'run_rooney.R')
source(R_file)
```

See `CF_problems.Rmd` for weird behaviour of conversion factors



```{r, warning=FALSE, echo=FALSE}
devtools::load_all(file.path('~/antigenic_space', 'libs', 'fasanalysis'))
plot_yield_rates_by_project()
```



## DEPRECATED Analysis 2: Pan cancer, non mut-context specific, conversion factors

2017-06-05 15:47 Interesting ideas to take from this analysis and maybe
implement in the future are project specific vs. pan can CFs and A2-negative vs.
A2-aspecific yield rates


Due to the oddly behaving conversion factors in the previous analysis, we made
another attempt, letting go of mutation spectrum specific conversion factors.
Additional settings are as follows:

   1. Results are shown with conversion factors computed on i) HLA-A\*02:01
      negative donors and all donors regardless of A\*02:01 status. Conversion
      factors are computed on both the pan-cancer and project specific levels
      and are not mutational spectrum specific.
   2. Indels are included in the $NS$ load and indel derived peptides are not
      excluded.
   3. Somatic mutations in driver genes are not excluded from analysis (other
      analyses of ours show no increased yield rate there).
   4. Expected neo-epitopes are tallied, as is most convenient when including
      indel derived peptides.


### Pan cancer conversion factors 

Using the global conversion factors, no apparent immune editing signal emerges.
Although the SKCM (melanoma), DLBC (B-cell lymphoma), ACC (adrenocortical
carcinoma) and CoMMpass (multiple myeloma) projects are observed to have a
sizeable amount of donors with less than expected neo-antigens, HLA typing is
either absent (DLBC, ACC, CoMMpass), not allowing to discern whether this is
a artefactual observation, or A\*02:01-allele presence does not influence the
observed $YR$, suggesting an artefactual observation (SKCM). Whether we compute
the conversion factors on donors that are A\*02:01 negative (left plot),
positive (middle plot) or for which HLA typing is currently available (right
plot), we get highly similar yet not *exactly* identical results as verified by
detailed inspection of the data results.

```{r, cache = F, eval = F, include = T, output = 'asis', warning=FALSE, echo=FALSE}
# Compute pan rates stratified on HLA-A\*02:01 positivity status and not stratified
pan_rates <- rbind(donor_summary[!project %in% blacklist_projs, mutcontext_ignore_rates(.SD),
                           by = c('hla_a0201_status')], 
   donor_summary[!project %in% blacklist_projs, mutcontext_ignore_rates(.SD)][, cbind(.SD, 'hla_a0201_status' = 'all')])


project_rates <- donor_summary[!project %in% blacklist_projs,
                               mutcontext_ignore_rates(.SD),
                               by = c('project', 'hla_a0201_status')]
project_rates <- rbind(donor_summary[!project %in% blacklist_projs,
                               mutcontext_ignore_rates(.SD),
                               by = c('project', 'hla_a0201_status')], 
                       donor_summary[!project %in% blacklist_projs,
                               mutcontext_ignore_rates(.SD),
                               by = c('project')][, cbind(.SD, 'hla_a0201_status' = 'all')]) 

mutcontext_ignore_devs <- donor_summary[,#!project %in% blacklist_projs,
       compute_deviations_epitopes(.SD, pan_rates), by = donor_id]
mutcontext_ignore_devs <- merge(mutcontext_ignore_devs,
                                donor_summary[, .(donor_id, project)],
                                by = 'donor_id')
saveRDS(mutcontext_ignore_devs, file.path(rds_dir, 'mutcontext_ignore_devs.rds'))

mutcontext_ignore_devs_bp <- donor_summary[,#!project %in% blacklist_projs,
       compute_deviations_epitopes(.SD, project_rates), by = donor_id]

mutcontext_ignore_devs_bp <- merge(mutcontext_ignore_devs_bp,
                                   donor_summary[, .(donor_id, project)],
                                   by = 'donor_id')
saveRDS(mutcontext_ignore_devs_bp,
        file.path(rds_dir, 'mutcontext_ignore_devs_bp.rds'))
```


```{r, eval = F, warning=FALSE, echo=FALSE}
project_rates[, table(hla_a0201_status)]
project_rates[, hla_a0201_status]
mutcontext_ignore_devs_bp[, table(hla_a0201_rate)]
mutcontext_ignore_devs_bp[, hla_a0201_rate]
mutcontext_ignore_devs[, table(hla_a0201_rate)]
```

```{r, warning=FALSE, echo=FALSE}
mutcontext_ignore_devs <- readRDS(file.path(rds_dir, 'mutcontext_ignore_devs.rds'))
mutcontext_ignore_devs_bp <- readRDS(file.path(rds_dir, 'mutcontext_ignore_devs_bp.rds'))
# cacher(mutcontext_ignore_devs <- donor_summary[!project %in% blacklist_projs,
#            compute_deviations_epitopes(.SD, project_rates), by = donor_id])
```


```{r, warning=FALSE, fig.width = 19, fig.height = 10}
devtools::load_all(file.path('~/antigenic_space', 'libs', 'fasanalysis'))
p1 <- plot_rooney_editing(mutcontext_ignore_devs[hla_a0201_rate == 'FALSE'],
                          y_val = 'B',
                          plot_title = 'CF computed pan-can on A*02:01 neg')
# p2 <- invisible(plot_rooney_editing(res[hla_a0201_rate == T],
#             y_val = 'B',
#             plot_title = 'Conversion factors computed on A*02:01 pos'))
p3 <- plot_rooney_editing(mutcontext_ignore_devs[hla_a0201_rate == 'all'], 
                          y_val = 'B',
                          plot_title = 'CF computed pan-can on independent of A:02 status')
plot_panel(list(p3, p1), ncol = 2)
```


```{r, eval = F, htmlcap = 'Medians of B values and equality tests (Wilcoxon rank sum) per project. Conversion factors computed on A:02 negative donors.', results = 'asis', warning=FALSE, echo=FALSE}
## Compare B-values between A\*02:01 pos and negative donors
wc_test <- mutcontext_ignore_devs_bp[hla_a0201_rate == F & !is.na(hla_a0201), {
  test <- wilcox.test(B ~ hla_a0201, data = .SD, alternative = 'g')
  # aggregate(B ~ hla_a0201, data = .SD, FUN = mean)$B
  # group_by(.SD, hla_a0201) %>% dplyr::summarise(mean(B, na.rm = T))
  .(pval = test$p.value,
    B_a0201_pos = median(.SD[hla_a0201, B], na.rm = T),
    B_a0201_neg = median(.SD[!hla_a0201, B], na.rm = T))
}, by = project]

wc_test[, 'p.adj' := p.adjust(pval, method = 'fdr')]
knitr::kable(wc_test)
# merge(res, wc_test, by = c('project', 'hla_a0201_rate'), allow.cartesian = T)
```


### Project specific conversion factors


As when using pan-cancer conversion factors, we see no differences between A\*02:01
positive and negative donors in any of the cancer projects. 


```{r, warning=FALSE, fig.width = 19, fig.height = 10}
p1 <- plot_rooney_editing(mutcontext_ignore_devs_bp[hla_a0201_rate == 'FALSE'],
                          y_val = 'B',
                          plot_title = 'CF computed by project on A*02:01 neg')
# p2 <- invisible(plot_rooney_editing(res[hla_a0201_rate == T],
#             y_val = 'B',
#             plot_title = 'Conversion factors computed on A*02:01 pos'))
p3 <- plot_rooney_editing(mutcontext_ignore_devs_bp[hla_a0201_rate == 'all'], y_val = 'B',
               plot_title = 'CF by project computed on\nindependent of A:02:01 status')
plot_panel(list(p3, p1), ncol = 2)
```

```{r, eval = F, warning=FALSE, echo=FALSE}
ggplot(data = mutcontext_ignore_devs_bp[hla_a0201_rate == 'all'], aes(x = B)) + facet_wrap(~project) +
  geom_histogram() + scale_x_log10()
qplot(R, data = mutcontext_ignore_devs_bp[project == 'THCA', ][hla_a0201_rate == 'all'])
qplot(R, data = mutcontext_ignore_devs_bp[project == 'CoMMpass', ][hla_a0201_rate == 'all'])
mutcontext_ignore_devs_bp[project == 'CoMMpass', ][hla_a0201_rate == 'all']
```


# Conclusion


The two analyses show no qualitative difference: neither point to the presence
of antigen loss in donors for which HLA matched predictions were done. Whether
to compute conversion factors on a per tumor or pan basis also does not make
a difference.


# MANA analysis


2017-05-11 14:47


```{r, eval = F, warning=FALSE, echo=FALSE}
plot_yield_rate_from_agg_dvc <- function(oname = 'dec_agg_dvc_aff', 
                                         title = '', w = .5) {
  agg_dvc <- w_readRDS(oname) 
  agg_dvc[, table(project_extended)]
  agg_dvc[, length(table(project))]

  p_dat <- agg_dvc[hla_a0201_status != 'Unknown',
     compute_prop_bool(length(pep_cont), sum(pep_cont, na.rm = T)), 
     by = .(project_extended, hla_a0201_status)]
  p_dat[, 'variable' := c('N', 'p', 'SE', 'N_neos')]
  mat <- as.data.table(tidyr::spread(p_dat, variable, V1))
  mat <- mat[N_neos != 0]

  p.vals <- agg_dvc[hla_a0201_status != 'Unknown', {
    tab <- table(pep_cont, hla_a0201_status)
    prop.test(tab[1:2, 1:2], alternative = 'less')$p.value
  }, by = project_extended]
  setnames(p.vals, 'V1', 'pval')

  p <- ggplot(mat, aes(x = project_extended, colour = hla_a0201_status, 
                       y = p, ymin = p - 1.96 * SE, ymax = p + 1.96 * SE)) + 
    geom_point(position = position_dodge(width = w)) + 
    geom_errorbar(position = position_dodge(width = w)) +
    geom_text(data = p.vals, colour = 'black', 
              mapping = aes(x = project_extended, y = mat[, max(p + 2.1 * SE)], 
                            hjust = 1,
                            label = sprintf("italic(p)==%.2f", pval)),
              inherit.aes = F,
              angle = 90, parse = TRUE) +
    scale_y_continuous(name = 'Fraction of mutations antigenic') + 
    scale_x_discrete(name = '') + 
    scale_colour_discrete(name = '') + 
    theme_fas(rotate_labels = 45, base_size = 12) +
    ggtitle(title)

  w_ggsave(sprintf('fraction_antigenic_%s', oname), w = 18, h = 16, units = 'cm')
  return(p)
}

plot_yield_rate_from_agg_dvc('agg_dvc',
  title = 'Including all genomic regions\nAll 4 filters')
```


```{r, eval = F, warning=FALSE, echo=FALSE}
plot_yield_rate_from_agg_dvc('de_clon_ccf_agg_dvc',
  title = 'Excluding subclonal, driver and essential gene mutations\nAll 4 filters')

plot_yield_rate_from_agg_dvc('ccf_agg_dvc',
  title = 'Including all genomic regions\nAll 4 filters')

plot_yield_rate_from_agg_dvc('de_clon_ccf_agg_dvc_aff',
  title = 'Excluding subclonal, driver and essential gene mutations\nAffinity filter')

plot_yield_rate_from_agg_dvc('de_clon_snv_ccf_agg_dvc',
  title = 'Excluding subclonal, driver and essential gene and non-SNV mutations\nFour filters')

plot_yield_rate_from_agg_dvc('de_clon_snv_ccf_agg_dvc_aff',
  title = 'Excluding subclonal, driver and essential gene and non-SNV mutations\nAffinity filter')
```
